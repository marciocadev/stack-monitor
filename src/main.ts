import { App, CfnOutput, CfnParameter, Stack, StackProps } from 'aws-cdk-lib';
import { LoggingLevel, SlackChannelConfiguration } from 'aws-cdk-lib/aws-chatbot';
import { RetentionDays } from 'aws-cdk-lib/aws-logs';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { Construct } from 'constructs';

export class MyStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    const topic = new Topic(this, 'StackMonitoringTopic', {
      topicName: 'stack-monitoring-topic',
    });
    new CfnOutput(this, 'TopicArnOutput', {
      value: topic.topicArn,
      exportName: 'stack-monitor-topic',
    });

    const workspaceId = new CfnParameter(this, 'Lazy::WorkspaceId');
    new CfnOutput(this, 'WorkspaceIdOutput', {
      value: workspaceId.valueAsString,
    });
    const channelId = new CfnParameter(this, 'Lazy::ChannelId');
    new CfnOutput(this, 'ChannelIdOutput', {
      value: channelId.valueAsString,
    });

    new SlackChannelConfiguration(this, 'StackMonitoringChatbot', {
      slackChannelConfigurationName: 'stack-monitoring-channel',
      slackWorkspaceId: workspaceId.valueAsString,
      slackChannelId: channelId.valueAsString,
      notificationTopics: [topic],
      logRetention: RetentionDays.ONE_DAY,
      loggingLevel: LoggingLevel.INFO,
    });

    console.log('teste1');
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new MyStack(app, 'stack-monitor-dev', { env: devEnv });
// new MyStack(app, 'stack-monitor-prod', { env: prodEnv });

app.synth();