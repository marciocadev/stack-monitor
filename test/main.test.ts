import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { MyStack } from '../src/main';

describe('Stack monitor', () => {
  let app:App, stack:MyStack, template:Template;

  beforeAll(() => {
    app = new App();
    stack = new MyStack(app, 'test');
    template = Template.fromStack(stack);
  });

  test('Snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

  test('Topic name', () => {
    template.hasResourceProperties('AWS::SNS::Topic', {
      TopicName: 'stack-monitoring-topic',
    });
  });

  test('SlackChannelConfiguration name', () => {
    template.hasResourceProperties('AWS::Chatbot::SlackChannelConfiguration', {
      ConfigurationName: 'stack-monitoring-channel',
    });
  });

  test('WorskpaceId type', () => {
    template.hasParameter('LazyWorkspaceId', {
      Type: 'String',
    });
  });

  test('ChannelId type', () => {
    template.hasParameter('LazyChannelId', {
      Type: 'String',
    });
  });
});