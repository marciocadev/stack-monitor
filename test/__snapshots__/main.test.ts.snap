// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack monitor Snapshot 1`] = `
Object {
  "Outputs": Object {
    "ChannelIdOutput": Object {
      "Value": Object {
        "Ref": "LazyChannelId",
      },
    },
    "WorkspaceIdOutput": Object {
      "Value": Object {
        "Ref": "LazyWorkspaceId",
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "LazyChannelId": Object {
      "Type": "String",
    },
    "LazyWorkspaceId": Object {
      "Type": "String",
    },
  },
  "Resources": Object {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "4cd567bf52e0f31674ae9e960efae8855b8b1249fa9997799ae47dc9f80e9ff8.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "StackMonitoringChatbotBC33B962": Object {
      "Properties": Object {
        "ConfigurationName": "stack-monitoring-channel",
        "IamRoleArn": Object {
          "Fn::GetAtt": Array [
            "StackMonitoringChatbotConfigurationRole7B350AFE",
            "Arn",
          ],
        },
        "LoggingLevel": "INFO",
        "SlackChannelId": Object {
          "Ref": "LazyChannelId",
        },
        "SlackWorkspaceId": Object {
          "Ref": "LazyWorkspaceId",
        },
        "SnsTopicArns": Array [
          Object {
            "Ref": "StackMonitoringTopic36E37431",
          },
        ],
      },
      "Type": "AWS::Chatbot::SlackChannelConfiguration",
    },
    "StackMonitoringChatbotConfigurationRole7B350AFE": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "chatbot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StackMonitoringChatbotLogRetention9D4D8D76": Object {
      "Properties": Object {
        "LogGroupName": "/aws/chatbot/stack-monitoring-channel",
        "LogGroupRegion": "us-east-1",
        "RetentionInDays": 1,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "StackMonitoringTopic36E37431": Object {
      "Properties": Object {
        "TopicName": "stack-monitoring-topic",
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
